puffin/src/scope_details.rs

```diff
@@ -0,0 +1,285 @@
+use std::{borrow::Cow, collections::HashMap, sync::Arc};
+
+use once_cell::sync::Lazy;
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use serde::{ser::SerializeStruct, Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::{clean_function_name, fetch_add_scope_id, short_file_name, ScopeId};
+
+// Scope details are stored separately from [`GlobalProfiler`] for the following reasons:
+//
+// 1. Almost every access only requires read access so there is no need for mutex lock the global profiler.
+// 2. Its important to guarantee that the profiler path is lock-free.
+// 2. External libraries like the http server or ui require read/write access to scopes.
+// But that can easily end up in deadlocks if a profile scope is executed while scope details are being read.
```

Comment:
I think the second second point here indicates the problem I've been trying to point out:

The scope _producer_ (processes using `profile_scope!` etc) need a global place to add scopes too.
But reusing this same global collection on the _consumer_ side (`puffin_egui`, `puffin_viewer`) seems like a mistake to me. It may very well be convenient (globals often are), but leads to things like the above mentioned deadlock.

Furthermore, any`profile_scope!` in `puffin_egui` or `puffin_viewer` will be writing over the same scopes.
`puffin_viewer` has a introspection mode where it profiles itself. This PR breaks it in two ways:
A) the scope info is wrong
B) it deadlocks after a while

![image](https://github.com/EmbarkStudios/puffin/assets/1148717/09e96ab4-9b80-4bda-a486-172fec8cfbe4)

On `main` all the data that `puffin_viewer` views is contained in a `FrameView`. It seems to me `FrameView` should contain a `ScopeCollection`.

Reply:
I managed to move it to `FrameView`, it comes with a little complexity for our users who insert custom scopes. They will have to use a `FrameView`, and make sure to get a full snapshot of scope details if the frame view was created after the scopes are registered.

Reply:
For convenience I can expose read access to the global scope collection from the `GlobalProfiler`, or would you like to keep that also restricted to `FrameView`. It is possible to do everything via a frame view